File: client/src/types/context.ts
export type ContextType =
  | "fact"
  | "preference"
  | "habit"
  | "health"
  | "work"
  | "relationship"
  | "goal_progress"
  | "conversation_summary"
  | "insight"
  | "other";

export type InspectedMemory = {
  id: string;
  title: string;
  context_type: ContextType;
  source: string;
  importance: number;
  similarity: number | null;
  embedded: boolean;
  score: number;
};

export type InspectedSignal = {
  title: string;
  source: string;
  source_ref: string | null;
  updated_at: string;
  summary: string | null;
};

export type ContextSection = {
  name: string;
  chars: number;
  tokens: number;
};

export type ContextInspection = {
  mode: string;
  channel: string;
  query: string;
  system_instruction: string;
  prompt: string;
  prompt_chars: number;
  token_estimate: number;
  sections: ContextSection[];
  memories: InspectedMemory[];
  live_signals: InspectedSignal[];
  history_count: number;
  notes: string[];
};

export type LongTermContextItem = {
  id: string;
  context_type: ContextType;
  title: string;
  content: string;
  summary: string | null;
  importance: number;
  confidence: number | null;
  source: string;
  source_ref: string | null;
  tags: string[];
  metadata: Record<string, unknown>;
  expires_at: string | null;
  is_archived: boolean;
  access_count: number;
  last_accessed_at: string | null;
  created_at: string;
  updated_at: string;
};

File: server/app/schemas/context.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.models.context import ContextSource, ContextType


class ContextCreateRequest(BaseModel):
    context_type: ContextType = "fact"
    title: str = Field(min_length=1, max_length=200)
    content: str = Field(min_length=1)
    summary: str | None = None
    importance: int = Field(default=5, ge=1, le=10)
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    source: ContextSource = "manual"
    source_ref: str | None = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    expires_at: datetime | None = None


class ContextUpdateRequest(BaseModel):
    context_type: ContextType | None = None
    title: str | None = Field(default=None, min_length=1, max_length=200)
    content: str | None = Field(default=None, min_length=1)
    summary: str | None = None
    importance: int | None = Field(default=None, ge=1, le=10)
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    tags: list[str] | None = None
    metadata: dict[str, Any] | None = None
    expires_at: datetime | None = None
    is_archived: bool | None = None


class ContextResponse(BaseModel):
    id: str
    context_type: ContextType
    title: str
    content: str
    summary: str | None
    importance: int
    confidence: float | None
    source: ContextSource
    source_ref: str | None
    tags: list[str]
    metadata: dict[str, Any]
    expires_at: datetime | None
    is_archived: bool
    access_count: int
    last_accessed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ContextInspectRequest(BaseModel):
    message: str = Field(default="", max_length=4000)
    mode: Literal["reactive", "proactive"] = "reactive"
    channel: Literal["dashboard", "whatsapp", "dev"] = "dashboard"


class InspectedMemory(BaseModel):
    id: str
    title: str
    context_type: ContextType
    source: ContextSource
    importance: int
    similarity: float | None
    embedded: bool
    score: float


class InspectedSignal(BaseModel):
    title: str
    source: ContextSource
    source_ref: str | None
    updated_at: datetime
    summary: str | None


class ContextSection(BaseModel):
    name: str
    chars: int
    tokens: int


class ContextInspectResponse(BaseModel):
    mode: str
    channel: str
    query: str
    system_instruction: str
    prompt: str
    prompt_chars: int
    token_estimate: int
    sections: list[ContextSection]
    memories: list[InspectedMemory]
    live_signals: list[InspectedSignal]
    history_count: int
    notes: list[str]
